src/lib.rs

```diff
@@ -1193,37 +1215,38 @@ impl<T: Counter> Histogram<T> {
     // Internal helpers
     // ********************************************************************************************
 
-    /// Compute the lowest (and therefore highest precision) bucket index that can represent the
-    /// value.
+    /// Compute the lowest (and therefore highest precision) bucket index whose sub-buckets can
+    /// represent the value.
     #[inline]
-    fn bucket_for(&self, value: u64) -> usize {
+    fn bucket_for(&self, value: u64) -> u8 {
         // Calculates the number of powers of two by which the value is greater than the biggest
         // value that fits in bucket 0. This is the bucket index since each successive bucket can
         // hold a value 2x greater. The mask maps small values to bucket 0.
-        (self.leading_zero_count_base - (value | self.sub_bucket_mask).leading_zeros()) as usize
+        self.leading_zero_count_base - (value | self.sub_bucket_mask).leading_zeros() as u8
     }
 
     #[inline]
-    /// Compute the position inside a bucket at which the given value should be recorded.
-    fn sub_bucket_for(&self, value: u64, bucket_index: usize) -> usize {
-        // For bucket_index 0, this is just value, so it may be anywhere in 0 to sub_bucket_count. For
-        // other bucket_index, this will always end up in the top half of sub_bucket_count: assume
-        // that for some bucket k > 0, this calculation will yield a value in the bottom half of 0
-        // to sub_bucket_count. Then, because of how buckets overlap, it would have also been in the
-        // top half of bucket k-1, and therefore would have returned k-1 in bucket_index_of(). Since
-        // we would then shift it one fewer bits here, it would be twice as big, and therefore in
-        // the top half of sub_bucket_count.
-        // TODO: >>> ?
-        (value >> (bucket_index + self.unit_magnitude)) as usize
+    /// Compute the position inside a bucket at which the given value should be recorded, indexed
+    /// from position 0 in the bucket (in the first half, which is not used past the first bucket).
+    /// For bucket_index > 0, the result will be in the top half of the bucket.
+    fn sub_bucket_for(&self, value: u64, bucket_index: u8) -> u32 {
+        // Since bucket_index is simply how many powers of 2 greater value is than what will fit in
+        // bucket 0 (that is, what will fit in [0, sub_bucket_count)), we shift off that many
+        // powers of two, and end up with a number in [0, sub_bucket_count).
+        // For bucket_index 0, this is just value. For bucket index k > 0, we know value won't fit
+        // in bucket (k - 1) by definition, so this calculation won't end up in the lower half of
+        // [0, sub_bucket_count) because that would mean it would also fit in bucket (k - 1).
```

Comment:
Yes! I've been thinking that at the end of this revisiting of the logic, it would be great to write up basically a design document. There was some talk of having it in a centralized place along with some test vectors in a dedicated repo or something, so I know Gil is open to the idea.

In particular I think it would be interesting to have both the slow version and optimized version of the core routines because it's often hard to reverse engineer the high level intent from all the bit wrangling.